File: app/src/main/java/com/stho/software/nyota/utilities/Angle.java
package com.stho.software.nyota.utilities;

/**
 * Created by shoedtke on 28.08.2016.
 */
public final class Angle {

    private Angle() {
        // Not instantiable
    }

    public enum AngleType {
        DEGREE_NORTH_EAST_SOUTH_WEST,
        AZIMUTH,
        ALTITUDE,
        LATITUDE,
        LONGITUDE,
    }

    public static float getAngleDifference(float x, float y) {
        return normalizeTo180(x - y);
    }

    public static double getAngleDifference(double x, double y) {
        return normalizeTo180(x - y);
    }

    public static float normalize(float angle) {
        while (angle > 360)
            angle -= 360;

        while (angle < 0)
            angle += 360;

        return angle;
    }

    public static double normalize(double angle) {
        angle = Math.IEEEremainder(angle, 360);

        if (angle < 0)
            angle += 360;

        return angle;
    }

    public static float normalizeTo180(float angle) {
        while (angle > 180)
            angle -= 360;

        while (angle <= -180)
            angle += 360;

        return angle;
    }

    public static double normalizeTo180(double angle) {
        angle = Math.IEEEremainder(angle, 360);

        if (angle < -180)
            angle += 360;

        if (angle > 180)
            angle -= 360;

        return angle;
    }

    public static String toString(double angle, AngleType angleType) {
         double alpha;
         switch (angleType) {

            default:
                alpha = Angle.normalizeTo180(angle);
                return sign(alpha) + Formatter.df2.format(Math.abs(alpha)) + "°";

            case AZIMUTH:
                alpha = Angle.normalize(angle);
                return Formatter.df0.format(Math.abs(alpha)) + "° " + northEastSouthWest(alpha);

            case ALTITUDE:
                alpha = Angle.normalizeTo180(angle);
                return sign(alpha) + Formatter.df0.format(Math.abs(alpha)) + "°";

            case LATITUDE:
                alpha = Angle.normalizeTo180(angle);
                return Formatter.df2.format(Math.abs(alpha)) + "° " + ((angle >= 0) ? "N" : "S");

            case LONGITUDE:
                alpha = Angle.normalizeTo180(angle);
                return Formatter.df2.format(Math.abs(alpha)) + "° " + ((angle >= 0) ? "E" : "W");

            case DEGREE_NORTH_EAST_SOUTH_WEST:
                alpha = Angle.normalize(angle);
                return Formatter.df0.format(Math.abs(alpha)) + "° " + northEastSouthWest(alpha);
        }
    }

    private static String sign(double x) { return (x < 0) ? "-" : "+"; }

    private static String northEastSouthWest(double angle) {
        final int n = (int)(((angle < 0) ? (angle + 282.5) : (angle + 22.5)) / 45);
        final String[] x = new String[]{ "N", "NE", "E", "SE", "S", "SW", "W", "NW", "N" };
        return x[n];
    }
}

File: app/src/main/java/com/stho/nyota/sky/utilities/Angle.kt
package com.stho.nyota.sky.utilities

import kotlin.math.abs

object Angle {

    fun toString(angle: Double, angleType: AngleType?): String =
        when (angleType) {
            AngleType.AZIMUTH -> {
                val alpha = normalize(angle)
                Formatter.df0.format(abs(alpha)) + "° " + northEastSouthWest(alpha)
            }
            AngleType.ALTITUDE -> {
                val alpha = normalizeTo180(angle)
                sign(alpha) + Formatter.df0.format(abs(alpha)) + "°"
            }
            AngleType.LATITUDE -> {
                val alpha = normalizeTo180(angle)
                Formatter.df2.format(abs(alpha)) + "° " + northSouth(alpha)
            }
            AngleType.LONGITUDE -> {
                val alpha = normalizeTo180(angle)
                Formatter.df2.format(abs(alpha)) + "° " + eastWest(alpha)
            }
            AngleType.DEGREE_NORTH_EAST_SOUTH_WEST -> {
                val alpha = normalize(angle)
                Formatter.df0.format(abs(alpha)) + "° " + northEastSouthWest(alpha)
            }
            AngleType.PITCH -> {
                val alpha = normalizeTo180(angle)
                sign(alpha) + Formatter.df0.format(abs(alpha)) + "°"
            }
            AngleType.ROLL -> {
                val alpha = normalizeTo180(angle)
                sign(alpha) + Formatter.df0.format(abs(alpha)) + "°"
            }
            else -> {
                val alpha = normalizeTo180(angle)
                sign(alpha) + Formatter.df2.format(abs(alpha)) + "°"
            }
        }

    fun toString(azimuth: Double, altitude: Double, angleType: AngleType?): String =
        when (angleType) {
            AngleType.ORIENTATION -> {
                toString(azimuth, AngleType.AZIMUTH) + " " + toString(altitude, AngleType.ALTITUDE)
            }
            else -> {
                val alpha = normalizeTo180(azimuth)
                val beta = normalizeTo180(altitude)
                sign(alpha) + Formatter.df2.format(abs(alpha)) + "°" + sign(beta) + Formatter.df2.format(abs(beta)) + "°"
            }
        }


    private fun sign(x: Double): String =
        if (x < 0) "-" else "+"

    private fun northSouth(angle: Double) =
        if (angle >= 0) "N" else "S"

    private fun eastWest(angle: Double) =
        if (angle >= 0) "E" else "W"

    /*
        angle in [0*45 - 22.5 , 1*45 - 22.5] -> DIRECTION[0] = "N"
        angle in [1*45 - 22.5 , 2*45 - 22.5] -> DIRECTION[1] = "NE"
        angle in [2*45 - 22.5 , 3*45 - 22.5] -> DIRECTION[3] = "E"
        ...
     */
    private fun northEastSouthWest(angle: Double): String =
        DIRECTION[((angle + 22.5) / 45.0).toInt()]

    private fun normalize(angle: Double): Double =
        Degree.normalize(angle)

    private fun normalizeTo180(angle: Double): Double =
        Degree.normalizeTo180(angle)

    enum class AngleType {
        DEGREE_NORTH_EAST_SOUTH_WEST, AZIMUTH, ALTITUDE, LATITUDE, LONGITUDE, PITCH, ROLL, ORIENTATION
    }

    private val DIRECTION = arrayOf("N", "NE", "E", "SE", "S", "SW", "W", "NW", "N")

}